dpdk/vr_dpdk_lcore.c

```diff
@@ -514,6 +650,48 @@ dpdk_lcore_exit()
     rte_free(lcore);
 }
 
+/* Handle an IPC command
+ * Returns -1 if if there is a stop command
+ */
+int
+vr_dpdk_lcore_cmd_handle(struct vr_dpdk_lcore *lcore)
+{
+    uint16_t cmd = (uint16_t)rte_atomic16_read(&lcore->lcore_cmd);
+    uint32_t cmd_param = (uint32_t)rte_atomic32_read(&lcore->lcore_cmd_param);
+    int ret = 0;
+    unsigned vif_idx;
+    struct vr_dpdk_rx_queue *rx_queue;
+    struct vr_dpdk_tx_queue *tx_queue;
+
+    switch (cmd) {
+    case VR_DPDK_LCORE_RX_RM_CMD:
+        vif_idx = cmd_param;
+        /* find the RX queue */
+        rx_queue = dpdk_lcore_rx_queue_find(lcore, vif_idx);
+        if (rx_queue) {
+            /* remove the queue from the lcore */
+            dpdk_lcore_rx_queue_remove(lcore, rx_queue);
+        }
+        break;
+    case VR_DPDK_LCORE_TX_RM_CMD:
+        vif_idx = cmd_param;
+        /* find the TX queue */
+        tx_queue = &lcore->lcore_tx_queues[vif_idx];
+        if (tx_queue->txq_queue_h) {
+            /* remove the queue from the lcore */
+            dpdk_lcore_tx_queue_remove(lcore, tx_queue);
+        }
+        break;
+    case VR_DPDK_LCORE_STOP_CMD:
+        ret = -1;
+        break;
+    }
+
+    rte_atomic16_set(&lcore->lcore_cmd, VR_DPDK_LCORE_NO_CMD);
```

Comment:
For the P6 family processors, locked operations (i.e. rte_atomic*) serialize all outstanding load and store operations (that is, wait for them to complete). This rule is also true for the Pentium 4 and Intel Xeon processors. Source: http://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-manual-325462.pdf